### ISSUE TITLE
Understanding How Masking Affects BCE Loss Function

### ISSUE THREAD

**Issue by chethanjjj** [tensorflow/tensorflow#50710]
Hi, I'm experimenting with masking and have come across an interesting observation. I've created a simple GRU-based model with zeroed-out weights and biases along with a simple X and Y tensor. Theoretically, because my weights and biases are set to zero I should get the same loss value for the same dataset regardless if masking is on or not. However, I find that masking affects my results and I'd like more clarification on why this is happening. Below is my code to reproduce the difference. Additionally, I am using version 2.4.1 and 2.4.0 for tensorflow and keras, respectively

```
import tensorflow as tf
from tensorflow.keras import layers
from tensorflow.keras import losses
from tensorflow.keras import regularizers
from tensorflow.keras import initializers
from tensorflow.keras import Model
from tensorflow.keras import backend as K
from tensorflow.keras.optimizers import Adam

GRU_SIZE = 50
DENSE_DIM = 200

# generate data
data = tf.random.uniform([2, 4, 4], maxval=2, dtype=tf.int32, seed=1)
data = tf.cast(data, tf.float32)
data = K.eval(tf.identity(data))

# zero-out specific timesteps in each patient
data[0, 2, :] = 0.0
data[0, 3, :] = 0.0
data[1, 2, :] = 0.0
data[1, 3, :] = 0.0

# use the 1st n-1 timesteps in X to predict the last n-1 timesteps in Y
x_dummy_data = data[:, 0:-1, :]
y_dummy_data = data[:, 1:, :]

# create the masked model
input_shape = (x_dummy_data.shape[1], x_dummy_data.shape[2])
inputs = layers.Input(shape=input_shape)
masking = layers.Masking(mask_value=0, input_shape=input_shape)
input_masked = masking.compute_mask(inputs)
output1 = layers.Dense(DENSE_DIM)(inputs)
output3 = layers.GRU(GRU_SIZE, return_sequences=True)(output1, mask=input_masked)
output4_code = layers.Dense(input_shape[1], 
                                                   activation='sigmoid', 
                                                   name="output_code")(output3)
model_mask = Model(inputs=inputs, outputs = output4_code)
model_mask.summary()
model_mask.compile(loss=tf.keras.losses.BinaryCrossentropy(reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE, from_logits=False),
                                     optimizer=Adam(learning_rate=1e-4))

# set weights and bias to zero
for ix, layer in enumerate(model_mask.layers):
     weights = layer.get_weights()
     for arr in weights:
        arr[(arr > 0) | (arr < 0)] = 0
     layer.set_weights(weights) 

# create the unmasked model
input_shape = (x_dummy_data.shape[1], x_dummy_data.shape[2])
inputs = layers.Input(shape=input_shape)
output1 = layers.Dense(DENSE_DIM)(inputs)
output3 = layers.GRU(GRU_SIZE, return_sequences=True)(output1)
output4_code = layers.Dense(input_shape[1], 
                                                   activation='sigmoid', 
                                                   name="output_code")(output3)
model_unmask = Model(inputs=inputs, outputs = output4_code)
model_unmask.summary()
model_unmask.compile(loss=tf.keras.losses.BinaryCrossentropy(reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE, from_logits=False),
                                     optimizer=Adam(learning_rate=1e-4))

# set weights and bias to zero
for ix, layer in enumerate(model_unmask.layers):
     weights = layer.get_weights()
     for arr in weights:
        arr[(arr > 0) | (arr < 0)] = 0
     layer.set_weights(weights) 

# evaluate on fake data
model_unmask.evaluate(x_dummy_data, y_dummy_data)
model_mask.evaluate(x_dummy_data, y_dummy_data)
```

The model_unmask.evaluate leads to a loss of 0.6931471824645996, while the masked model is 0.4620981216430664. Can you explain how the latter is calculated? If weights and biases are set to 0 and the last layer is sigmoid activated then the only input into the loss function is a tensor of 0.5's. Shouldn't these both be the same when you take an average over the batch?

**Comment by tilakrayal** [comment]
@chethanjjj ,

I ran the code shared and face a different error, please find the [gist](https://colab.research.google.com/gist/tilakrayal/4fe74ef9ad6fb5d7efa8b43e8da045fc/untitled50710.ipynb) here and share all dependencies to replicate the issue or share a colab gist with the reported error.Also, please let us know the tensorflow version you are using.Thanks!

**Comment by chethanjjj** [comment]
Got you. I fixed it by adding the dependencies, addressing any remaining errors, and indicating the dependency versions in the first paragraph.

**Comment by tilakrayal** [comment]
@ymodak  ,
I was able to reproduce the issue in tf v2.4,v2.5 and nightly.Please find the gist [here](https://colab.research.google.com/gist/tilakrayal/546ad1618d21f1bec70625996b04f3cb/untitld50710.ipynb).

**Comment by ymodak** [comment]
This question is better asked on [TensorFlow Forum](https://discuss.tensorflow.org/) since it is not a bug or feature request. There is also a larger community that reads questions there.
Thanks!

**Comment by chethanjjj** [comment]
Hmmm that's a valid place to ask too, but how is this not a bug? Theoretically, both should give the same value if all weights and biases are zeroed out and you were able to reproduce it.

**Comment by ymodak** [comment]
I have cross posted your issue to keras repo for further review from keras team. Please follow the tagged thread for updates. Thanks!
